src/util-macset.c

```diff
@@ -0,0 +1,398 @@
+                   we need to insert the current item. We do this backwards,
+                   since we expect the latest item to match more likely than
+                   the first */
+                for (int i = ms->last[side] - 1; i >= 0; i--) {
+                    uint8_t *addr2 = (uint8_t*) ((ms->buf[side]) + i);
+                    /* If we find a match, we return early with no action */
+                    if (likely(memcmp(addr2, addr, sizeof(MacAddr)) == 0)) {
+                        return;
+                    }
+                }
+            }
+            /* Otherwise, we insert the new address at the end */
+            memcpy(ms->buf[side] + ms->last[side], addr, sizeof(MacAddr));
+            ms->last[side]++;
+            break;
+    }
+}
+
+void MacSetAdd(MacSet *ms, uint8_t *src_addr, uint8_t *dst_addr)
+{
+    if (ms == NULL)
+        return;
+    MacUpdateEntry(ms, src_addr, MAC_SET_SRC);
+    MacUpdateEntry(ms, dst_addr, MAC_SET_DST);
+}
+
+static inline int MacSetIterateSide(MacSet *ms, MacSetIteratorFunc IterFunc,
+                                    MacSetSide side, void *data)
+{
+    int ret = 0;
+    switch (ms->state[side]) {
+        case EMPTY_SET:
+            return 0;
+        case SINGLE_MAC:
+            ret = IterFunc((uint8_t*) ms->singles[side], side, data);
+            if (unlikely(ret != 0)) {
+                return ret;
+            }
+            break;
+        case MULTI_MAC:
+            for (int i = 0; i < ms->last[side]; i++) {
+                ret = IterFunc((uint8_t*) ms->buf[side][i], side, data);
+                if (unlikely(ret != 0)) {
+                    return ret;
+                }
+            }
+            break;
+    }
+    return 0;
+}
+
+int MacSetForEach(MacSet *ms, MacSetIteratorFunc IterFunc, void *data)
+{
+    int ret = 0;
+    if (ms == NULL)
+        return 0;
+
+    ret = MacSetIterateSide(ms, IterFunc, MAC_SET_SRC, data);
+    if (ret != 0) {
+        return ret;
+    }
+    return MacSetIterateSide(ms, IterFunc, MAC_SET_DST, data);
+}
+
+int MacSetSize(MacSet *ms)
+{
+    int size = 0;
+    if (ms == NULL)
+        return 0;
+
+    switch(ms->state[MAC_SET_SRC]) {
+        case EMPTY_SET:
+            /* pass */
+            break;
+        case SINGLE_MAC:
+            size += 1;
+            break;
+        case MULTI_MAC:
+            size += ms->last[MAC_SET_SRC];
+            break;
+    }
+    switch(ms->state[MAC_SET_DST]) {
+        case EMPTY_SET:
+            /* pass */
+            break;
+        case SINGLE_MAC:
+            size += 1;
+            break;
+        case MULTI_MAC:
+            size += ms->last[MAC_SET_DST];
+            break;
+    }
+    return size;
+}
+
+void MacSetReset(MacSet *ms)
+{
+    if (ms == NULL)
+        return;
+    ms->state[MAC_SET_SRC] = ms->state[MAC_SET_DST] = EMPTY_SET;
+    ms->last[MAC_SET_SRC] = ms->last[MAC_SET_DST] = 0;
+}
+
+void MacSetFree(MacSet *ms)
```

Comment:
Since all macset's are created with `MacSetInit(10)`, I suggest that one call to `SC_ATOMIC_SUB` be made with the total size being returned.

Reply:
We could also make this number configurable?

Reply:
What would be the advantage of doing that?

Reply:
It would allow to tune memory usage against expected variability of MAC addresses. I assume that it it unlikely to have more than 10 MAC addresses per side and flow, but if you know that there will never be more than _n_ then you would be able to configure Suricata accordingly. Just an idea.

Apart from that, I am not sure I understand what you mean. I need to consider both `ms->buf`s because they are allocated independently of each other, so I can only add/subtract their memory usage when I know they are in use. Also, they are only allocated when a second MAC address per flow is seen for the first time, so I need to actually check whether I need to subtract the allocated bytes from the global memory usage. I agree that this can be calculated locally within `MacSetFree` and only one call to `SC_ATOMIC_SUB` can be made. Is that what you meant?

Reply:
It might be nice to be able to track (in a stats entry perhaps) what the max number of MAC addresses per side is in some real world setups. Then we can adjust the default accordingly.

Reply:
Sure. I can introduce a counter for that.

Reply:
Done, will have `decoder.max_mac_addrs_src` and `decoder.max_mac_addrs_dst` in next PR.